stan/math/rev/functor/algebra_solver_powell.hpp

```diff
@@ -177,6 +52,118 @@ Eigen::VectorXd algebra_solver_powell(
+  check_nonnegative("algebra_solver_powell", "function_tolerance",
+                    function_tolerance);
+  check_positive("algebra_solver_powell", "max_num_steps", max_num_steps);
+
+  // Construct the Powell solver
+  auto f_wrt_x = [&](const auto& x) {
+    return apply([&](const auto&... args) { return f(x, msgs, args...); },
+                 args_vals_tuple);
+  };
+
+  hybrj_functor_solver<decltype(f_wrt_x)> fx(f_wrt_x);
+  Eigen::HybridNonLinearSolver<decltype(fx)> solver(fx);
+
+  // Check dimension unknowns equals dimension of system output
+  check_matching_sizes("algebra_solver", "the algebraic system's output",
+                       fx.get_value(x_val), "the vector of unknowns, x,", x);
+
+  // Solve the system
+  return algebra_solver_powell_call_solver_(solver, fx, x_val, 0,
+                                            relative_tolerance,
+                                            function_tolerance, max_num_steps);
+}
+
+/** Implementation of autodiff powell solver. */
+template <typename F, typename T, typename... T_Args,
+          require_eigen_vector_t<T>* = nullptr,
+          require_any_st_var<T_Args...>* = nullptr>
+Eigen::Matrix<var, Eigen::Dynamic, 1> algebra_solver_powell_impl(
+    const F& f, const T& x, std::ostream* msgs, double relative_tolerance,
+    double function_tolerance, int64_t max_num_steps,
+    const T_Args&... args) {  // NOLINT(runtime/int)
+  const auto& x_val = to_ref(value_of(x));
+  auto arena_args_tuple = std::make_tuple(to_arena(args)...);
+  auto args_vals_tuple = std::make_tuple(eval(value_of(args))...);
+
+  check_nonzero_size("algebra_solver_powell", "initial guess", x_val);
+  check_finite("algebra_solver_powell", "initial guess", x_val);
+  check_nonnegative("alegbra_solver_powell", "relative_tolerance",
+                    relative_tolerance);
+  check_nonnegative("algebra_solver_powell", "function_tolerance",
+                    function_tolerance);
+  check_positive("algebra_solver_powell", "max_num_steps", max_num_steps);
+
+  // Construct the Powell solver
+  auto f_wrt_x = [&](const auto& x) {
+    return apply([&](const auto&... args) { return f(x, msgs, args...); },
+                 args_vals_tuple);
+  };
+
+  hybrj_functor_solver<decltype(f_wrt_x)> fx(f_wrt_x);
+  Eigen::HybridNonLinearSolver<decltype(fx)> solver(fx);
+
+  // Check dimension unknowns equals dimension of system output
+  check_matching_sizes("algebra_solver", "the algebraic system's output",
+                       fx.get_value(x_val), "the vector of unknowns, x,", x);
+
+  // Solve the system
+  Eigen::VectorXd theta_dbl = algebra_solver_powell_call_solver_(
+      solver, fx, x_val, 0, relative_tolerance, function_tolerance,
+      max_num_steps);
+
+  Eigen::MatrixXd Jf_x;
+  Eigen::VectorXd f_x;
+
+  jacobian(f_wrt_x, theta_dbl, f_x, Jf_x);
+
+  using ret_type = Eigen::Matrix<var, Eigen::Dynamic, -1>;
+  auto Jf_xT_lu_ptr = make_chainable_ptr(Jf_x.transpose().fullPivHouseholderQr()); // Lu
```

Comment:
@jgaeb if I replace the fullPivLu with this householder, things work. Was the LU there previously a partial pivot or a full pivot one? If it was a full pivot LU, then Eigen says they're comparable in the table here: https://eigen.tuxfamily.org/dox/group__TutorialLinearAlgebra.html, so I assume they are comparable enough for our purposes. You could check your benchmarks as well to make sure the householder is not way slower or something.

As far as the LU goes, I agree something is funky. If I change this line to:

```cpp
auto Jf_xT_lu_ptr = make_chainable_ptr(Jf_x.transpose().fullPivLu());
```

then it's like you say the code breaks because it seems the `plain_type_t` has converted the LU decomposition into a Matrix. The QR does not do this. I'm not sure the issue. But I guess first do we need LU?

Reply:
It's a partial pivot—we can expect the matrices to be invertible, which means we only need partial pivoting, which substantially speeds up the calculation. (QR is, I think, about 2x slower than full pivoting LU anyway, too, unfortunately.) If we leave it as QR, it'll end up being slower than the old version, unfortunately.

@SteveBronder — Do you have any thoughts on how to address this? In principle, it shouldn't be hard: We just need to stash (1) an `n x n` matrix where the L and U factors are stored, and (2) the permutation matrix generated by the column pivots. Both of these are members of a `PartialPivLU` matrix, but I haven't been able to think of a way of storing them without copying them, and `Eigen` doesn't provide a constructor for `PartialPivLU` from an LU decomposition and permutation matrix, so we'd have to reimplement the solve method in-line. It's not super complicated, but it seems a little ugly / hacky. Is there some way we can just create a wrapper class or something that will fix the `plain_type_t` conversion?